lib/parsers/cr2_parser.rb

```diff
@@ -0,0 +1,119 @@
+class FormatParser::CR2Parser
+  include FormatParser::IOUtils
+
+  # Variables
+  TIFF_HEADER = [0x49, 0x49, 0x2a, 0x00]
+  CR2_HEADER  = [0x43, 0x52, 0x02, 0x00]
+
+  PREVIEW_ORIENTATION_TAG = 0x0112
+  PREVIEW_RESOLUTION_TAG = 0x011a
+  PREVIEW_IMAGE_OFFSET_TAG = 0x0111
+  PREVIEW_IMAGE_BYTE_COUNT_TAG = 0x0117
+
+  def call(io)
+    io = FormatParser::IOConstraint.new(io)
+
+    tiff_header = safe_read(io, 8)
+
+    # Check whether it's a CR2 file
+    tiff_bytes = tiff_header[0..3].bytes
+    magic_bytes = safe_read(io, 4).unpack('C4')
+
+    return if !magic_bytes.eql?(CR2_HEADER) || !tiff_bytes.eql?(TIFF_HEADER)
+
+    # Offset to IFD #0 where the preview image data is located
+    # For more information about CR2 format,
+    # see http://lclevy.free.fr/cr2/
+    # and https://github.com/lclevy/libcraw2/blob/master/docs/cr2_poster.pdf
+    if0_offset = tiff_header[4..7].reverse.bytes.collect { |c| c.to_s(16) }.join.hex
+
+    set_orientation(io, if0_offset)
+    set_resolution(io, if0_offset)
+    set_preview(io, if0_offset)
+
+    # Check CanonAFInfo or CanonAFInfo2 tags in maker notes for width & height
+    exif_offset = parse_ifd(io, if0_offset, 0x8769)
+    makernote_offset = parse_ifd(io, exif_offset[0], 0x927c)
+    af_info = parse_ifd(io, makernote_offset[0], 0x0026)
+
+    # Old Canon models have CanonAFInfo tags (0x0012)
+    # Newer models have CanonAFInfo2 tags (0x0026) instead
+    # See https://sno.phy.queensu.ca/~phil/exiftool/TagNames/Canon.html
+    if !af_info.nil?
+      parse_new_model(io, af_info[0], af_info[1])
+    else
+      af_info = parse_ifd(io, makernote_offset[0], 0x0012)
```

Comment:
There are constants for some offsets/tag IDs, but not for others, is that intentional?

Reply:
All tags moved to constants.